source-postgres/replication.go

```diff
@@ -0,0 +1,348 @@
+			//      on these same edge cases, with the exception of TRUNCATE/TYPE/ORIGIN.
+			//      (See https://github.com/debezium/debezium/blob/52333596dec168a22674e04f8cd94b1f56607f73/debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/pgoutput/PgOutputMessageDecoder.java#L106)
+			//
+			//   3. Force a full refresh of the relevant table. This will re-scan the
+			//      current table contents and then begin replication after the event
+			//      that caused this to fail.
+			//
+			//      But of course if this wasn't a one-off event it will fail again the
+			//      next time a problematic statement is executed on this table.
+			return errors.Errorf("unhandled message (Type=%v, WALStart=%q)", msg.Type(), xld.WALStart)
+		}
+	}
+}
+
+func (s *ReplicationStream) processChange(ctx context.Context, eventType string, xld pglogrepl.XLogData, tuple *pglogrepl.TupleData, relID uint32, handler ChangeEventHandler) error {
+	if !s.inTransaction {
+		return errors.Errorf("got %s message without a transaction in progress", eventType)
+	}
+
+	rel, ok := s.relations[relID]
+	if !ok {
+		return errors.Errorf("unknown relation ID %d", relID)
+	}
+
+	fields := make(map[string]interface{})
+	if tuple != nil {
+		for idx, col := range tuple.Columns {
+			colName := rel.Columns[idx].Name
+			switch col.DataType {
+			case 'n':
+				fields[colName] = nil
+			case 't':
+				val, err := s.decodeTextColumnData(col.Data, rel.Columns[idx].DataType)
+				if err != nil {
+					return errors.Wrap(err, "error decoding column data")
+				}
+				fields[colName] = val
+			default:
+				return errors.Errorf("unhandled column data type %v", col.DataType)
+			}
+		}
+	}
+
+	evt := &ChangeEvent{
+		Type:      eventType,
+		LSN:       xld.WALStart,
+		Namespace: rel.Namespace,
+		Table:     rel.RelationName,
+		Fields:    fields,
+	}
+	evt.XID.Set(s.transactionXID)
+	if err := handler(evt); err != nil {
+		return errors.Wrap(err, "error handling change event")
+	}
+	return nil
+}
+
+func (s *ReplicationStream) decodeTextColumnData(data []byte, dataType uint32) (interface{}, error) {
+	var decoder pgtype.TextDecoder
+	if dt, ok := s.connInfo.DataTypeForOID(dataType); ok {
+		decoder, ok = dt.Value.(pgtype.TextDecoder)
+		if !ok {
+			decoder = &pgtype.GenericText{}
+		}
+	} else {
+		decoder = &pgtype.GenericText{}
+	}
+	if err := decoder.DecodeText(s.connInfo, data); err != nil {
+		return nil, err
+	}
+	return decoder.(pgtype.Value).Get(), nil
+}
+
+// receiveXLogData returns the next transaction log message from the database,
+// blocking until a message is available, the context is cancelled, or an error
+// occurs. In the process it takes care of sending Standby Status Update messages
+// back to the database, so it must be called regularly over the life of a
+// replication stream.
+func (s *ReplicationStream) receiveXLogData(ctx context.Context) (pglogrepl.XLogData, error) {
+	for {
+		select {
```

Comment:
Github ate the context for this comment (I should have used reviewable), but as I recall it was an explicit check for context cancellation which is unlikely to ever trip, because the cancellation is far more likely to be observed after that check and while sitting in the following ReceiveMessage call instead.

Basically you still need to inspect errors coming from ReceiveMessage _anyway_ to determine if they're due to cancellation or something else, so the ahead-of-time check doesn't really help.

Reply:
I see. The logic is a bit convoluted, but the way I'd explain it is that we're not checking ahead-of-time, the check of context-doneness there is asking "why am I at the top of the loop this time around?".

The `ReceiveMessage` call can time out either because the top-level context was cancelled or because the "time to send a standby update" deadline passed, and to me it seemed cleaner to just check both of those conditions at the start of the loop.